File: tasks/it290925.py
# Создам код, который вычисляет оценки по L2 (среднеквадратичная) и по L1 (сумма модулей),
# строит предсказания hat_y2 (L2) и hat_y1 (L1), считает значения J1 и J2 в найденных точках,
# и показывает графики: данные + обе линии аппроксимации и функции потерь J1(b), J2(b).
import numpy as np
import matplotlib.pyplot as plt
from statistics import median

# Данные
x = np.array([1, 2, 3, 4, 5, 6], dtype=float)
y = np.array([1, 1.8, 2.6, 3.4, 6, 4], dtype=float)
log_x = np.log(x)
n = len(y)

# --- Оценка L2 (нормальная регрессия) ---
# Модель: y = a + b * log_x
X = np.vstack([np.ones(n), log_x]).T
theta_l2, residuals, rank, s = np.linalg.lstsq(X, y, rcond=None)
a_l2, b_l2 = theta_l2[0], theta_l2[1]

# Предсказания и значения функций потерь в theta_hat_L2
yhat_l2 = a_l2 + b_l2 * log_x
J2_at_l2 = np.sum((y - yhat_l2)**2)          # L2 loss (sum of squared errors)
J1_at_l2 = np.sum(np.abs(y - yhat_l2))       # L1 loss at L2 estimate

# --- Поиск оценки L1 (минимизация суммы абсолютных отклонений) ---
# Для фиксированного b оптимальный a — медиана(y - b*log_x).
# Перебор по b: грубый сет, потом уточнение.
def evaluate_Js_for_b(b_vals):
    a_vals = np.array([median((y - b * log_x).tolist()) for b in b_vals])
    residuals = y.reshape(-1,1) - (a_vals + b_vals * log_x.reshape(-1,1))
    J1_vals = np.sum(np.abs(residuals), axis=0)
    J2_vals = np.sum(residuals**2, axis=0)
    return a_vals, J1_vals, J2_vals

# Грубый поиск
b_grid1 = np.linspace(-5, 5, 4001)
a_grid1, J1_grid1, J2_grid1 = evaluate_Js_for_b(b_grid1)
idx1 = np.argmin(J1_grid1)
b_est1 = b_grid1[idx1]
a_est1 = a_grid1[idx1]
# Уточним локально два раза
for span, pts in [(0.5,2001), (0.05,2001)]:
    b_grid_local = np.linspace(b_est1-span, b_est1+span, pts)
    a_grid_local, J1_grid_local, J2_grid_local = evaluate_Js_for_b(b_grid_local)
    idx_local = np.argmin(J1_grid_local)
    b_est1 = b_grid_local[idx_local]
    a_est1 = a_grid_local[idx_local]

a_l1, b_l1 = a_est1, b_est1
yhat_l1 = a_l1 + b_l1 * log_x
J1_at_l1 = np.sum(np.abs(y - yhat_l1))
J2_at_l1 = np.sum((y - yhat_l1)**2)

# Также подготовим J1(b) и J2(b) в широком диапазоне для отображения графиков
b_plot = np.linspace(b_grid1[0], b_grid1[-1], 8001)
a_plot = np.array([median((y - b * log_x).tolist()) for b in b_plot])
res_plot = y.reshape(-1,1) - (a_plot + b_plot * log_x.reshape(-1,1))
J1_plot = np.sum(np.abs(res_plot), axis=0)
# Для J2: оптимальный a_for_b (минимизирует SSE) это mean(y - b*log_x)
a_plot_mean = np.array([np.mean(y - b * log_x) for b in b_plot])
res_plot2 = y.reshape(-1,1) - (a_plot_mean + b_plot * log_x.reshape(-1,1))
J2_plot = np.sum(res_plot2**2, axis=0)

# --- Вывод числовых результатов ---
print("Оценки L2 (минимизация суммы квадратов): a = {:.6f}, b = {:.6f}".format(a_l2, b_l2))
print("J2(theta_hat_L2) = {:.6f}".format(J2_at_l2))
print("J1(theta_hat_L2) = {:.6f}".format(J1_at_l2))
print()
print("Оценки L1 (минимизация суммы модулей): a = {:.6f}, b = {:.6f}".format(a_l1, b_l1))
print("J1(theta_hat_L1) = {:.6f}".format(J1_at_l1))
print("J2(theta_hat_L1) = {:.6f}".format(J2_at_l1))
print()
print("hat_y (L2):", np.round(yhat_l2, 6))
print("hat_y (L1):", np.round(yhat_l1, 6))

# --- График 1: данные + обе аппроксимации ---
plt.figure(figsize=(9,6))
plt.scatter(x, y)
x_smooth = np.linspace(x.min(), x.max(), 200)
plt.plot(x_smooth, a_l2 + b_l2 * np.log(x_smooth), linewidth=2, label='L2 fit (least squares)')
plt.plot(x_smooth, a_l1 + b_l1 * np.log(x_smooth), linestyle='--', linewidth=2, label='L1 fit (least absolute deviations)')
plt.legend()
plt.title('Data and fits: L2 vs L1')
plt.xlabel('x'); plt.ylabel('y')
plt.grid(True, alpha=0.3)
plt.show()

# # --- График 2: J1(b) и J2(b) как функции b (с оптимальным a для каждого b) ---
# plt.figure(figsize=(9,6))
# plt.plot(b_plot, J1_plot, linewidth=2, label='J1(b) = sum |y - (a_med(b) + b ln x)|')
# plt.plot(b_plot, J2_plot, linewidth=2, label='J2(b) = min_a sum (y - (a + b ln x))^2 (a = mean residual)')
# # Отметим найденные точки
# plt.scatter([b_l1], [J1_at_l1], s=80)
# plt.scatter([b_l2], [J2_at_l2], s=80)
# plt.axvline(b_l1, linestyle=':', linewidth=1)
# plt.axvline(b_l2, linestyle=':', linewidth=1)
# plt.legend()
# plt.title('Loss curves J1(b) и J2(b) (оптимальный a для каждого b)')
# plt.xlabel('b'); plt.ylabel('Loss')
# plt.grid(True, alpha=0.3)
# plt.show()

# Чтобы пользователь видел числа в удобном виде — создадим словарь результатов
results = {
    'L2_theta': (a_l2, b_l2),
    'J2_at_L2': float(J2_at_l2),
    'J1_at_L2': float(J1_at_l2),
    'L1_theta': (a_l1, b_l1),
    'J1_at_L1': float(J1_at_l1),
    'J2_at_L1': float(J2_at_l1),
    'yhat_L2': yhat_l2.tolist(),
    'yhat_L1': yhat_l1.tolist(),
}
results

# Остатки
res_L2 = y - yhat_l2
res_L1 = y - yhat_l1

# Локальные L2 и L1 нормы (нарастающие)
cum_L2 = [np.linalg.norm(res_L2[:i+1], 2) for i in range(len(res_L2))]
cum_L1 = [np.linalg.norm(res_L1[:i+1], 1) for i in range(len(res_L1))]

# Построение графика
plt.figure(figsize=(9, 6))
plt.plot(x, cum_L2, "o-", label="||Y - Ŷ_L2||_2")
plt.plot(x, cum_L1, "s--", label="||Y - Ŷ_L1||_1")
plt.xlabel("x")
plt.ylabel("Накопленная ошибка")
plt.title("Накопленные нормы ошибок L1 и L2")
plt.legend()
plt.grid(alpha=0.3)
plt.show()

File: tasks/it290925.m
%% QR
% Создаем вектор значений t
t = linspace(0, 2*pi, 24);

% Формируем матрицу, где каждый столбец = вектор v для t(k)
A = zeros(length(t), 4);
for k = 1:length(t)
    A(k, :) = [sin(t(k)), cos(t(k)), sin(t(k) + pi/4), 1];
end

% Применяем QR-разложение
[Q, R, E] = qr(A);

% Q — ортонормированная матрица
% R — верхнетреугольная матрица
E
R
A*E-Q*R

%% Ridge
X = [1, 2, 3, 4, 5, 6];
det(X'*X)
% y = [1, 1.8, 2.6, 3.4, 6, 4];
y = [1, 1.8, 2.4, 2.8, 3.0, 3.1];

lambda_values = 0:0.1:10;
J = zeros(size(lambda_values)); % для хранения функции потерь

for i = 1:length(lambda_values)
    lambda_2 = lambda_values(i);
    % Ridge regression
    hat_theta = inv((X'*X + lambda_2*eye(size(X,2)))) * (X'*y);
    % MSE
    J(i) = (1/2) * sum((X*hat_theta - y).^2);
end

figure;
plot(lambda_values, J, 'LineWidth', 2);
xlabel('lambda_2');
ylabel('J(theta estimated)');
title('Функция потерь J(theta estimated) от lambda_2');
grid on;

%% 17:34

% t=(0:0.1:10)';
t=(0:0.1:1000)';
% X = [sin(t) cos(t) ones(size(t))];
% X=[sin(t) exp(t/100)];
% X=[sin(t) exp(-t/100)];
X = [sin(t) cos(t)];
% X = [sin(t) t/1000]
% X = [sin(t) ones(size(t))];
% theta=[1;2;3];
theta=[1;2];
Y0=X*theta;
% Y=Y0+rand(length(t), 1);
Y=Y0+rand(length(t), 1)-0.5;
% S=rand(length(t), 1)-0.5;
% S=rand(length(t), 1);
% Y=Y0+S.*sin(t);
hat_theta=inv(X'*X)*X'*Y
E=Y-X*hat_theta;
mean(E)
corr(E,X(:,1))
corr(E,X(:,2))

% возможно выполнялись для помехи (если выполнились)
% точно нет (если не выполнились)
% стат тест -?- дисперсия постоянная
% наличие трендов. вероятность роста больше чем убывания при росте.
% вероятность падения больше чем убывания если падает. корреляция.

%% new
t=(0:0.1:2*pi)';

a = sin(t) + rand(length(t), 1);
b = sin(t) .* rand(length(t), 1);

var_a = var(a)
var_b = var(b)

figure;

subplot(2,1,1)
% autocorr(a)
autocorr(a, 'NumLags', length(a)-1)
title('Автокорреляция a')

subplot(2,1,2)
% autocorr(b)
autocorr(b, 'NumLags', length(a)-1)
title('Автокорреляция b')

